### PR TITLE
feat(draw): add generic draw image helpers

### LINKED ISSUE #4995: Make most of the draw_sw_image logic reusable
### Introduce the problem

We have just added support for [Renesas Dave2D](https://github.com/lvgl/lvgl/tree/master/src/draw/renesas/dave2d) and I found that [lv_draw_dave2d_image.c](https://github.com/lvgl/lvgl/blob/master/src/draw/renesas/dave2d/lv_draw_dave2d_image.c) is mostly the duplicate of [lv_draw_sw_img.c](https://github.com/lvgl/lvgl/blob/master/src/draw/sw/lv_draw_sw_img.c). This way it's hard to follow that changes and fixes on all the places. 

I believe this problem applies to other GPU implementation too.

### Examples and cases
See above

### Suggested solution
In [lv_draw_sw_img.c](https://github.com/lvgl/lvgl/blob/master/src/draw/sw/lv_draw_sw_img.c) the only SW render specif part is [img_draw_core()](https://github.com/lvgl/lvgl/blob/master/src/draw/sw/lv_draw_sw_img.c#L288-L505). The rest is about decoding and tiling. 

I think we should make `img_draw_normal` and `img_draw_tiled` public and add a callback parameter which is something like `img_draw_core` or its platform specific equivalent.

What do you think?

cc @XuNeo @nicusorcitu @jeremy-baker

### PR DESCRIPTION
**Pull request by kisvegabor**
### Description of the feature or fix

Fixes #4995 

Nothing new, just made the static functions global.

### Checkpoints
- [ ] Run `code-format.py` from the scripts folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [ ] Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed
- [ ] Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant. 
- [ ] Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- [ ] If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/release/v8.3/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/release/v8.3/Kconfig).
 
Be sure the following conventions are followed:
- [ ] Follow the [Styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [ ] Prefer `enum`s instead of macros. If inevitable to use `define`s export them with `LV_EXPORT_CONST_INT(defined_value)` right after the `define`.
- [ ] In function arguments prefer `type name[]` declaration for array parameters instead of `type * name`
- [ ] Use typed pointers instead of `void *` pointers
- [ ] Do not `malloc` into a static or global variables. Instead declare the variable in `lv_global_t` structure in [`lv_global.h`](https://github.com/lvgl/lvgl/blob/master/src/core/lv_global.h) and mark the variable with `(LV_GLOBAL_DEFAULT()->variable)` when it's used. See a detailed description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#memory-management).
- [ ] Widget constructor must follow the `lv_<widget_name>_create(lv_obj_t * parent)` pattern.
- [ ] Widget members function must start with `lv_<module_name>` and should receive `lv_obj_t *` as first argument which is a pointer to widget object itself.  
- [ ] `struct`s should be used via an API and not modified directly via their elements.
- [ ] `struct` APIs should follow the widgets' conventions. That is to receive a pointer to the `struct` as the first argument, and the prefix of the `struct` name should be used as the prefix of the function name too (e.g.  `lv_disp_set_default(lv_disp_t * disp)`)
- [ ] Functions and `struct`s which are not part of the public API must begin with underscore in order to mark them as "private".
- [ ] Arguments must be named in H files too.
- [ ] To register and use callbacks one of the following needs to be followed (see a detailed description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#callbacks)):
  - For both the registration function and the callback pass a pointer to a `struct` as the first argument. The `struct` must contain `void * user_data` field.
  - The last argument of the registration function must be `void * user_data` and the same `user_data` needs to be passed as the last argument of the callback.
  - Callback types not following these conventions should end with `xcb_t`.

### CODE REVIEW

**Comment by GorgonMeducer** [comment]
It looks good for me.
This PR opens a lot of new opportunities.

**Comment by XuNeo** [comment]
cc @FASTSHIFT

**Comment by XuNeo** [comment]
It looks good for general GPUs without too many limitations. For VG-Lite, it does need some customization. See implementation from FASTSHIFT. 

https://github.com/FASTSHIFT/lvgl/blob/feat_draw_vg_lite/src/draw/vg_lite/lv_draw_vg_lite_img.c

**Comment by kisvegabor** [comment]
Cool! I also think that it's a cheap way to integrate simpler GPUs, for VGLite level we need something more custom for sure. 

So finally do you agree to merge it?

File: src/draw/lv_draw_image.c [hunk 1]
```diff
@@ -116,6 +121,131 @@ lv_image_src_t lv_image_src_get_type(const void * src)
...
+                                const lv_area_t * img_area, const lv_area_t * clipped_img_area,
+                                lv_draw_image_core_cb draw_core_cb)
+{
+    lv_draw_image_sup_t sup;
+    sup.alpha_color = draw_dsc->recolor;
+    sup.palette = decoder_dsc->palette;
+    sup.palette_size = decoder_dsc->palette_size;
+
+    /*The whole image is available, just draw it*/
+    if(decoder_dsc->decoded || decoder_dsc->img_data) {
```
Comment (XuNeo) [inline src/draw/lv_draw_image.c:221]: I'll update PR https://github.com/lvgl/lvgl/pull/4948 in these days and check if any better method can be used to detect available image areas.

File: src/draw/lv_draw_image.c [hunk 2]
```diff
@@ -116,6 +121,131 @@ lv_image_src_t lv_image_src_get_type(const void * src)
...
+        draw_core_cb(draw_unit, draw_dsc, decoder_dsc, &sup, img_area, clipped_img_area);
+    }
+    /*Draw in smaller pieces*/
+    else {
+        lv_area_t relative_full_area_to_decode = *clipped_img_area;
+        lv_area_move(&relative_full_area_to_decode, -img_area->x1, -img_area->y1);
+
+        lv_area_t relative_decoded_area;
```
Comment (XuNeo) [inline src/draw/lv_draw_image.c:229]: This should be removed too in future.

I mean this part.
```c
        relative_decoded_area.x1 = LV_COORD_MIN;
        relative_decoded_area.y1 = LV_COORD_MIN;
        relative_decoded_area.x2 = LV_COORD_MIN;
        relative_decoded_area.y2 = LV_COORD_MIN;
        lv_result_t res = LV_RESULT_OK;
```

**Review by XuNeo — APPROVED** [review]
It looks good to me.

**Comment by kisvegabor** [comment]
Okay, I merge it now and we can fine tune it later.
